src/duration/mod.rs

```diff
@@ -0,0 +1,356 @@
+
+        let mut chars = input.char_indices().peekable();
+        let mut end_index = 0;
+
+        // First loop through all numeric characters
+        while let Some((i, _)) = chars.next_if(|(_, c)| char::is_numeric(*c)) {
+            end_index = i + 1;
+        }
+
+        // Parse the numeric characters as a u128
+        let value = if end_index != 0 {
+            s[0..end_index].parse::<u128>()?
+        } else {
+            return Err(DurationFragmentParseError::ExpectedNumber);
+        };
+
+        // Loop through all alphabetic characters
+        let start_index = end_index;
+        while let Some((i, _)) = chars.next_if(|(_, c)| char::is_alphabetic(*c)) {
+            end_index = i + 1;
+        }
+
+        // Parse the alphabetic characters as a supported duration unit
+        let unit = if end_index != start_index {
+            s[start_index..end_index]
+                .parse::<DurationUnit>()
+                .map_err(|_| DurationFragmentParseError::UnitParseError)?
+        } else {
+            return Err(DurationFragmentParseError::ExpectedCharacter);
+        };
+
+        // If there are characters left which are not alphabetic, we return an
+        // error
+        if chars.peek().is_some() {
+            return Err(DurationFragmentParseError::ExpectedCharacter);
+        }
+
+        Ok(Self { value, unit })
+    }
+}
+
+impl Display for DurationFragment {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}{}", self.value, self.unit)
+    }
+}
+
+impl DurationFragment {
+    /// Returns the amount of whole milliseconds encoded by this
+    /// [`DurationFragment`].
+    pub fn millis(&self) -> u128 {
+        self.value * self.unit.millis()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use rstest::rstest;
+    use serde::{Deserialize, Serialize};
+
+    #[rstest]
+    #[case("15d 2m 2s", 1296122)]
+    #[case("1h", 3600)]
+    #[case("1m", 60)]
+    #[case("1s", 1)]
+    fn parse(#[case] input: &str, #[case] output: u64) {
+        let dur: Duration = input.parse().unwrap();
+        assert_eq!(dur.as_secs(), output);
+    }
+
+    #[rstest]
+    #[case(
+        "2d2",
+        DurationParseError::FragmentError(DurationFragmentParseError::ExpectedCharacter)
+    )]
+    #[case(
+        "-1y",
+        DurationParseError::FragmentError(DurationFragmentParseError::ExpectedNumber)
+    )]
+    #[case(
+        "1D",
+        DurationParseError::FragmentError(DurationFragmentParseError::UnitParseError)
+    )]
+    #[case("1ä", DurationParseError::InvalidInput)]
+    #[case(" ", DurationParseError::InvalidInput)]
+    fn parse_invalid(#[case] input: &str, #[case] expected_err: DurationParseError) {
+        let err = Duration::from_str(input).unwrap_err();
+        assert_eq!(err, expected_err)
+    }
+
+    #[rstest]
+    #[case("15d 2m 2s")]
+    #[case("1h 20m")]
+    #[case("1m")]
+    #[case("1s")]
+    fn to_string(#[case] duration: &str) {
```

Comment:
This only tests already-canonical durations, also need to consider whether `70s` should stay that or become `1m10s` (for example).

Reply:
I will add tests for these cases.

Reply:
Added in 9e9257d